server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java

```diff
@@ -371,6 +401,19 @@ public long getInterruptedQueryCount()
     return interruptedQueryCount.get();
   }
 
+  private static byte[] getRequestContent(HttpServletRequest request) throws IOException
+  {
+    int contentSize = request.getContentLength();
+    byte[] content = new byte[contentSize];
+    int bytesRead = request.getInputStream().read(content);
```

Comment:
This is not the right way to read from an InputStream. See the spec for the `read(byte[])` method: it may do a partial read; it's not guaranteed to read to the end of the stream, or the array, or really to any particular number of bytes.

`ByteStreams.toByteArray` is a simple way to read an input stream fully into a byte array. Check out its code to see what it's doing.

Reply:
Thanks, good catch.

I ended up having the objectMapper deserialize the content from the request directly, like the handling for non-SQL Query objects below this block.

ByteStreams.toByteArray reads until the stream returns -1, which I think sets it to EOF state, which breaks later on here:

```
2017-10-20T03:34:31,113 WARN [qtp378797968-75] org.eclipse.jetty.server.HttpChannel - //localhost:8888/druid/v2/sql/avatica/
java.lang.IllegalStateException: State STREAM != EOF
	at org.eclipse.jetty.server.HttpInput.setReadListener(HttpInput.java:602) ~[jetty-server-9.3.19.v20170502.jar:9.3.19.v20170502]
	at org.eclipse.jetty.proxy.AsyncProxyServlet.proxyRequestContent(AsyncProxyServlet.java:58) ~[jetty-proxy-9.3.19.v20170502.jar:9.3.19.v20170502]
	at org.eclipse.jetty.proxy.ProxyServlet.service(ProxyServlet.java:115) ~[jetty-proxy-9.3.19.v20170502.jar:9.3.19.v20170502]
	at io.druid.server.AsyncQueryForwardingServlet.service(AsyncQueryForwardingServlet.java:268) ~[druid-server-0.11.1-SNAPSHOT.jar:0.11.1-SNAPSHOT]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) ~[javax.servlet-api-3.1.0.jar:3.1.0]
...
```